dolosse/generic_json_consumer.py

```diff
@@ -0,0 +1,153 @@
+        except (ValueError, TypeError):
+            valid = False
+        if not valid:
+            try:
+                following = int(decode[other_key], 16) #Catch hexadecimals
+                valid = True
+            except (ValueError, TypeError):
+                valid = False
+        if isinstance(decode[other_key], dict):
+            populate(variables,
+                     decode[other_key],
+                     count,
+                     name+": ")
+        else:
+            if name in variables:
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(
+                        variables[name][-1])
+            else:
+                variables[name] = [0]*(count-1)
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(0)
+                    #Remember, the list might still be empty here
+
+
+def get_args():
+    """Creates and returns a kafka consumer"""
+    parser = ArgumentParser()
+    parser.add_argument("--topic", default="default",
+                        help="The name of the kafka topic from which to read")
+    parser.add_argument("--group", default="my-group",
+                        help="The kafka group in which the topic exists")
+    parser.add_argument("--server", default="localhost:9092",
+                        help="Name and port of the kafka server (e.g. localhost:9092)")
+    arguments = parser.parse_args()
+
+    # To consume latest messages and auto-commit offsets
+    return arguments.topic, arguments.group, arguments.server
+
+def redraw(figures):
+    """Redraw all figures from a dict of figures"""
+    for key in figures:
+        figures[key].canvas.draw()
+        figures[key].canvas.flush_events()
+
+
+def update_data(data, value):
+    """Update the list of data, given some json input"""
+    try:
+        update = loads(value)
+    except decoder.JSONDecodeError as problem:
+        print("Error: Flawed JSON in value: ",
+              value)
+        print("Error message: ", problem)
+        update = ""
+    data.update(update)
+
+
+def graph(topic, group, server):
+    """Main function - reads json data from a kafka topic and graphs it"""
+
+    screen = initscr()
+    cbreak()
+    screen.nodelay(True)
+    screen.getch()
+
+    print("Generic Graphing Consumer")
+    print("Press any key to exit...")
+
+    data = {}
+    line = {}
+    plots = {}
+    figures = {}
+    axes = {}
+    count = 0
+    pyplot.ion() #Interactive mode on
+    configuration = {"bootstrap.servers": server,
+                     "group.id": group,
+                     "default.topic.config" : {
+                         "auto.offset.reset": "earliest"},
+                     "enable.auto.commit": False}
+    consumer = Consumer(configuration)
+    consumer.subscribe([topic])
+    keyboard_input = ERR
```

Comment:
Why instantiate to `ERR` and not `None`?

Reply:
The default return value from the screen.getch() function (if the user has not yet pressed a key) is ERR. I therefore read the value of that function and then quit as soon as the keyboard_input value is not ERR; thus, it seems sensible to instantiate to ERR, so that I only need to check against one possible value whether or not to continue.

Reply:
Gotcha. This makes a stronger case for using None. You could simple set to None and then check when it becomes not none. Same thing and you're not implying an error happens when it doesn't.

Reply:
It will become not None that very first time that I call screen.getch() (which returns a value of ERR if there is no input), and will therefore exit after a single loop whether the user has pressed a key or not.